app/services/input_type/base.rb

```diff
@@ -0,0 +1,101 @@
+module InputType
+  class Base
+    class Resolver
+      attr_reader :input, :scope
+
+      def initialize(input, scope)
+        @input = input
+        @scope = scope
+      end
+
+      def preview
+        ready? ? resolved_value : preview_value
+      end
+
+      def value
+        ready? ? resolved_value : raise(TemplateInput::ValueNotReady, "Input '#{@input.name}' is not ready for rendering")
+      end
+
+      def preview_value
+        "$#{@input.input_type.upcase}_INPUT[#{@input.name}]"
+      end
+
+      # should be defined in descendants
+      def ready?
+        raise NotImplementedError
+      end
+
+      # should be defined in descendants
+      def resolved_value
+        raise NotImplementedError
+      end
+    end
+
+    # ----- OWN METHODS ----
+
+    def self.input_type_name
+      name.split('::').last.sub('Input', '').underscore
+    end
+
+    def self.humanized_name
+      input_type_name.humanize
```

Comment:
This is fallback if the InputProvider forget to implement the method, in such case, there is low probablity it would get found, so it would probably not get translated anyway, right? But I don't mind including it, it might get lucky :)